### PR TITLE
[ASP-2210] Harden License Manage data model

### PR DESCRIPTION
**Pull request by julianaklulo**
#### What
Rewrite the API to use a normalized database.

#### Why
To improve the usability of the API.

`Task`: https://jira.scania.com/browse/ASP-2210

---

#### Peer Review
Please follow the upstream omnivector documentation concerning
[peer-review guidelines](https://github.com/omnivector-solutions/Documentation/blob/main/Contributing/pr_review_standards.md#peer-review).

### CODE REVIEW

File: backend/Dockerfile [hunk 1]
```diff
@@ -5,14 +5,14 @@ WORKDIR /app
...
-    POETRY_HOME=/opt/poetry POETRY_VERSION=1.1.13 python && \
+    POETRY_HOME=/opt/poetry POETRY_VERSION=1.4.0 python && \
     cd /usr/local/bin && \
     ln -s /opt/poetry/bin/poetry && \
     poetry config virtualenvs.create false
 
 COPY ./pyproject.toml ./poetry.lock* /app/
 
-RUN poetry install --no-root --no-dev
+RUN poetry install --no-root
```
Comment (dusktreader) [inline backend/Dockerfile]: Why do we need the dev dependencies in the Docker images?

If we need this for our CI pipeline, then we should create a separate version (Dockerfile.ci) that includes these and that is used by the docker-compose.yaml used in the CI. Our docker images in prod should not include dev dependencies.

File: backend/alembic/env.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
+
 from logging.config import fileConfig
 
 from alembic import context
 from sqlalchemy import engine_from_config, pool
 from sqlalchemy_utils import ScalarListType
 
-from lm_backend.config import settings
-from lm_backend.table_schemas import metadata as lm_backend_metadata
+sys.path = ['', '..'] + sys.path[1:]
```
Comment (dusktreader) [inline backend/alembic/env.py]: Why do we have this here?

File: backend/alembic/versions/0003--implement_new_data_struture_schema.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+# revision identifiers, used by Alembic.
+revision = "594b52f2c088"
+down_revision = "5308f26e8b60"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (dusktreader) [inline backend/alembic/versions/0003--implement_new_data_struture_schema.py]: These comments should be removed after the auto-generated migrations are reviewed by the dev.

File: backend/alembic/versions/0003--implement_new_data_struture_schema.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    op.drop_table("license")
+    op.drop_table("booking")
+    op.drop_index("ix_config_client_id", table_name="config")
+    op.drop_table("config")
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table(
```
Comment (dusktreader) [inline backend/alembic/versions/0003--implement_new_data_struture_schema.py]: I don't think we should have downgrade commands here. This will be a breaking change. We should not ever go back once we apply this.

File: backend/docker-compose.yaml [hunk 5]
```diff
@@ -30,3 +30,52 @@ services:
...
+      context: .
+      dockerfile: ./Dockerfile
+    environment:
+      DATABASE_URL: postgresql://omnivector:rats@pgsql:5432/ratsdb
+      ARMASEC_DOMAIN: ${ARMASEC_DOMAIN}
+      ARMASEC_AUDIENCE: ${ARMASEC_AUDIENCE}
+      ARMASEC_DEBUG: ${ARMASEC_DEBUG}
+      LOG_LEVEL: ${LOG_LEVEL}
+    command: >
+      sh -c  "/usr/local/bin/pip install alembic==1.8.1 psycopg2==2.9.5 && alembic -c alembic/alembic.ini upgrade head"
```
Comment (dusktreader) [inline backend/docker-compose.yaml]: Why doesn't this just use poetry? We should not be including hard-coded dependency specifications in a compose file.

File: backend/lm_backend/api/__init__.py [hunk 6]
```diff
@@ -1,10 +1,20 @@
...
-api_v1.include_router(router_config, prefix="/config", tags=["Config"])
+api = APIRouter()
+api.include_router(router_clusters, prefix="/clusters", tags=["Cluster"])
+api.include_router(router_configurations, prefix="/configurations", tags=["Configuration"])
+api.include_router(router_license_servers, prefix="/license_servers", tags=["License Server"])
+api.include_router(router_products, prefix="/products", tags=["Product"])
+api.include_router(router_features, prefix="/features", tags=["Feature"])
+api.include_router(router_inventories, prefix="/inventories", tags=["Inventory"])
+api.include_router(router_jobs, prefix="/jobs", tags=["Job"])
+api.include_router(router_bookings, prefix="/bookings", tags=["Booking"])
```
Comment (dusktreader) [inline backend/lm_backend/api/__init__.py:20]: Beautiful!!

File: backend/lm_backend/api/cruds/booking.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from lm_backend.api.cruds.generic import CreateSchemaType, GenericCRUD, ModelType, UpdateSchemaType
+from lm_backend.api.models import Booking, Feature, Inventory
+from lm_backend.api.schemas import BookingCreateSchema
+
+
+class BookingCRUD(GenericCRUD):
+    def __init__(self, model: ModelType, create_schema: CreateSchemaType, update_schema: UpdateSchemaType):
+        super().__init__(model, create_schema, update_schema)
```
Comment (dusktreader) [inline backend/lm_backend/api/cruds/booking.py]: This `__init__()` method looks like it's unnecessary.

File: backend/lm_backend/api/cruds/booking.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+    async def create(self, db_session: AsyncSession, obj=BookingCreateSchema) -> ModelType:
+        """Create a new booking.
+
+        Use a somewhat complex logic to check if a booking can be made and insert it in one query. In this
+        case, implement the `INSERT columns FROM select_query WHERE EXISTS (SELECT subquery)` SQL paradigm
+        using SQLAlchemy query API. A  booking will only be created if there are enough licenses available
+        to make the booking.
```
Comment (dusktreader) [inline backend/lm_backend/api/cruds/booking.py:24]: Good docstring!

File: backend/lm_backend/api/cruds/booking.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            .select_from(Feature)
+            .join(Booking, and_(Feature.id == Booking.feature_id, Booking.feature_id == obj.feature_id))
+            .join(Inventory, Feature.id == Inventory.feature_id)
+            .group_by(Feature.id)
+            .having(
+                func.sum(Booking.quantity)
+                + func.max(Inventory.used)
+                + obj.quantity
+                + func.max(Feature.reserved)
+                <= func.max(Inventory.total)
```
Comment (dusktreader) [inline backend/lm_backend/api/cruds/booking.py]: I think it would be good to write this inequality in the docstring in a clearer way if possible. This isn't terribly friendly to read in the sense of an inequality expression. It's easy to read as code, but I think it would be good if the core logic of this method was clearly expressed above.

File: backend/lm_backend/api/cruds/generic.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        self.create_schema = create_schema
+        self.update_schema = update_schema
+
+    async def create(self, db_session: AsyncSession, obj: CreateSchemaType) -> ModelType:
+        """Creates a new object in the database."""
+        db_obj = self.model(**obj.dict())
+        try:
+            async with db_session.begin():
+                db_session.add(db_obj)
+            await db_session.commit()
```
Comment (dusktreader) [inline backend/lm_backend/api/cruds/generic.py]: Is an explicit commit needed here? I thought you got an implicit commit when you use `begin()` as a context manager.

File: backend/lm_backend/api/cruds/generic.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+                        setattr(db_obj, field, value)
+                await db_session.flush()
+            except Exception as e:
+                logger.error(e)
+                raise HTTPException(status_code=400, detail="Object could not be updated.")
+
+            await db_session.refresh(db_obj)
+        return db_obj
+
+    async def delete(self, db_session: AsyncSession, id: int) -> bool:
```
Comment (dusktreader) [inline backend/lm_backend/api/cruds/generic.py]: Why is this function signature returning a bool? There's no return expressions in here.

File: backend/lm_backend/api/models.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
+"""
+Database models for the database tables in the License Manager API.
+Tables:
+    - license_servers
```
Comment (dusktreader) [inline backend/lm_backend/api/models.py]: Now that we have so many models, we really need to break this file up. I think maybe we should instead have a `models/` folder with a separate module for each one. Then the `__init__.py` module can import each one so that abstraction doesn't impact how other modules import them. What do you think?

File: backend/lm_backend/api/models.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+    features = relationship(
+        "Feature", back_populates="configurations", lazy="selectin", cascade="all, delete-orphan"
+    )
+
+    searchable_fields = [name]
+    sortable_fields = [name, cluster_id]
+
+    def __repr__(self):
+        return (
+            f"Config(id={self.id}, "
```
Comment (dusktreader) [inline backend/lm_backend/api/models.py]: The open paren after `Config` is never closed.

File: backend/lm_backend/api/routes/bookings.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+from lm_backend.api.models import Booking
+from lm_backend.api.schemas import BookingCreateSchema, BookingSchema, BookingUpdateSchema
+from lm_backend.permissions import Permissions
+from lm_backend.security import guard
+from lm_backend.session import get_session
+
+router = APIRouter()
+
+
+crud_booking = BookingCRUD(Booking, BookingCreateSchema, BookingUpdateSchema)
```
Comment (dusktreader) [inline backend/lm_backend/api/routes/bookings.py:16]: I really like this approach for the crud endpoints. Wrapping the logic for mapping from user-supplied data to database entries and back is very nice. It makes the routers so clean for basic CRUD operations.

File: backend/lm_backend/api/routes/bookings.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    "/",
+    response_model=BookingSchema,
+    status_code=status.HTTP_201_CREATED,
+    dependencies=[Depends(guard.lockdown(Permissions.BOOKING_EDIT))],
+)
+async def create_booking(
+    booking: BookingCreateSchema,
+    db_session: AsyncSession = Depends(get_session),
+):
+    """Create a new booking."""
```
Comment (dusktreader) [inline backend/lm_backend/api/routes/bookings.py:29]: It might be nice to add some logging in the routes so that we can see request details when they come in.

File: backend/lm_backend/api/routes/clusters.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+
+@router.delete(
+    "/{cluster_id}",
+    status_code=status.HTTP_200_OK,
+    dependencies=[Depends(guard.lockdown(Permissions.CLUSTER_EDIT))],
+)
+async def delete_cluster(cluster_id: int, db_session: AsyncSession = Depends(get_session)):
+    """
+    Delete a cluster from the database.
```
Comment (dusktreader) [inline backend/lm_backend/api/routes/clusters.py:104]: For multi-line docstrings there should be an empty line after the first one.

File: backend/lm_backend/api/routes/jobs.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+@router.delete(
+    "/slurm_job_id/{slurm_job_id}/cluster/{cluster_id}",
+    status_code=status.HTTP_200_OK,
+    dependencies=[Depends(guard.lockdown(Permissions.JOB_EDIT))],
+)
+async def delete_job_by_slurm_id(
+    slurm_job_id: str, cluster_id: int, db_session: AsyncSession = Depends(get_session)
+):
+    """
+    Delete a job from the database and associated bookings.
```
Comment (dusktreader) [inline backend/lm_backend/api/routes/jobs.py:82]: Need a blank line after the first one for multi-line docstrings.

File: backend/lm_backend/api/routes/jobs.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+    Since the slurm_job_id can be the same across clusters, we need the cluster_id to validate.
+    """
+    jobs = await crud_job.read_all(db_session=db_session, search=slurm_job_id)
+
+    for job in jobs:
+        if job.cluster_id == cluster_id:
+            await crud_job.delete(db_session=db_session, id=job.id)
+            return {"message": "Job deleted successfully"}
+
+    raise HTTPException(status_code=404, detail="The job doesn't exist in this cluster.")
```
Comment (dusktreader) [inline backend/lm_backend/api/routes/jobs.py:94]: Might be clearer to raise this exception if the `jobs` list is empty.

File: backend/lm_backend/api/routes/jobs.py [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+    "/slurm_job_id/{slurm_job_id}/cluster/{cluster_id}",
+    status_code=status.HTTP_200_OK,
+    dependencies=[Depends(guard.lockdown(Permissions.JOB_VIEW))],
+)
+async def read_job_by_slurm_id(
+    slurm_job_id: str, cluster_id: int, db_session: AsyncSession = Depends(get_session)
+):
+    """
+    Read a job from the database and associated bookings.
+    Uses the slurm_job_id and the cluster_id to filter the job.
```
Comment (dusktreader) [inline backend/lm_backend/api/routes/jobs.py:108]: blank line after first in multiline docstring.

File: backend/lm_backend/api/routes/products.py [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+@router.delete(
+    "/{product_id}",
+    status_code=status.HTTP_200_OK,
+    dependencies=[Depends(guard.lockdown(Permissions.PRODUCT_EDIT))],
+)
+async def delete_product(product_id: int, db_session: AsyncSession = Depends(get_session)):
+    """Delete a product from the database and associated features."""
+    await crud_product.delete(db_session=db_session, id=product_id)
+    return {"message": "Product deleted successfully"}
```
Comment (dusktreader) [inline backend/lm_backend/api/routes/products.py]: I wonder if you could make the GenericCRUD return a message like this with the name of the SQLAlchemy model in the string so you don't have to repeat this in every router delete.

File: backend/lm_backend/api/schemas.py [hunk 21]
```diff
@@ -0,0 +1,363 @@
+"""
+Schemas for the License Manager API.
```
Comment (dusktreader) [inline backend/lm_backend/api/schemas.py]: I wonder if it would be good to break the schemas up into a module for each of the different entities so they aren't all in one huge file.

File: backend/lm_backend/api/schemas.py [hunk 22]
```diff
@@ -0,0 +1,363 @@
...
+
+class BaseUpdateSchema(BaseModel):
+    """
+    Base class for updating objects in the database.
+    """
+
+    pass
+
+
+class LicenseServerType(str, Enum):
```
Comment (dusktreader) [inline backend/lm_backend/api/schemas.py]: This should probably go into a "constants.py" module or something similar since it isn't itself a schema.

File: backend/lm_backend/api/schemas.py [hunk 23]
```diff
@@ -0,0 +1,363 @@
...
+    License server to be created in the database.
+    """
+
+    config_id: int
+    host: str
+    port: int
+    type: LicenseServerType
+
+    @validator("port")
+    def port_must_be_positive(cls, v):
```
Comment (dusktreader) [inline backend/lm_backend/api/schemas.py]: I think you could use a constrained type from Pydantic for this: https://docs.pydantic.dev/latest/usage/types/#constrained-types

File: backend/lm_backend/database.py [hunk 24]
```diff
@@ -1,31 +1,22 @@
 """
-Persistent data storage for the backend
+Persistent data storage for the API.
```
Comment (dusktreader) [inline backend/lm_backend/database.py:2]: If we are going to rename "backend" to "API", it might make sense to change the subproject name as well? Maybe that's too much to take on right now.

File: backend/lm_backend/database.py [hunk 25]
```diff
@@ -1,31 +1,22 @@
...
-    metadata.create_all(engine)
+Base = declarative_base()
 
 
 def render_sql(query) -> str:
     """
     Render a sqlalchemy query into a string for debugging.
     """
-    return query.compile(dialect=database._backend._dialect, compile_kwargs={"literal_binds": True})
+    return query.compile(compile_kwargs={"literal_binds": True})
```
Comment (dusktreader) [inline backend/lm_backend/database.py:19]: Isn't the dialect needed to properly render the query args?

File: backend/tests/api/routes/conftest.py [hunk 26]
```diff
@@ -0,0 +1,325 @@
...
+            "name": "converge_super",
+            "product_id": product_id,
+            "config_id": configuration_id,
+            "reserved": 0,
+        },
+    ]
+
+    inserted_features = await insert_objects(features_to_add, Feature)
+
+    # inventories_to_add = [
```
Comment (dusktreader) [inline backend/tests/api/routes/conftest.py]: Should this just be removed?

File: backend/tests/api/routes/conftest.py [hunk 27]
```diff
@@ -0,0 +1,325 @@
...
+            "name": "abaqus",
+            "product_id": product_id,
+            "config_id": configuration_id,
+            "reserved": 0,
+        },
+    ]
+
+    inserted_feature = await insert_objects(feature_to_add, Feature)
+
+    # inventory_to_add = [{
```
Comment (dusktreader) [inline backend/tests/api/routes/conftest.py]: Remove?

File: backend/tests/api/routes/test_clusters.py [hunk 28]
```diff
@@ -0,0 +1,170 @@
...
+
+
+@mark.asyncio
+async def test_add_cluster__success(
+    backend_client: AsyncClient,
+    inject_security_header,
+    read_object,
+    clean_up_database,
+):
+    """
```
Comment (dusktreader) [inline backend/tests/api/routes/test_clusters.py]: We should be consistent with docstrings on the test functions. I personally dont't think they are necessary. We should just be consistent. This docstring doesn't really add much.

File: backend/tests/api/routes/test_bookings.py [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+    assert response.status_code == 200
+
+    response_booking = response.json()
+    assert response_booking["job_id"] == create_one_booking[0].job_id
+    assert response_booking["feature_id"] == create_one_booking[0].feature_id
+    assert response_booking["quantity"] == create_one_booking[0].quantity
+
+
+@mark.asyncio
+async def test_delete_booking__success(
```
Comment (dusktreader) [inline backend/tests/api/routes/test_bookings.py:176]: I think we need a test case for failures on each endpoint where parameters can cause issues. So, get by id, delete, update, etc.

File: backend/tests/conftest.py [hunk 30]
```diff
@@ -1,65 +1,181 @@
...
-    Make sure our database is empty at the end of each test
+    A test app to be used by the test backend_client.
     """
-    yield
 
-    count = await database.fetch_all("SELECT COUNT(*) FROM license")
-    assert count[0][0] == 0
+    from lm_backend.main import app as test_app
+
+    yield test_app
```
Comment (dusktreader) [inline backend/tests/conftest.py:52]: Why is this needed? Can't the test modules just import app from main.py themselves?

File: backend/tests/conftest.py [hunk 31]
```diff
@@ -1,65 +1,181 @@
...
+    async with get_session() as sess:
+        await sess.execute(sqlalchemy.text(f"""DELETE FROM {models.Booking.__tablename__};"""))
+        await sess.execute(sqlalchemy.text(f"""DELETE FROM {models.Job.__tablename__};"""))
+        await sess.execute(sqlalchemy.text(f"""DELETE FROM {models.Inventory.__tablename__};"""))
+        await sess.execute(sqlalchemy.text(f"""DELETE FROM {models.Feature.__tablename__};"""))
+        await sess.execute(sqlalchemy.text(f"""DELETE FROM {models.Product.__tablename__};"""))
+        await sess.execute(sqlalchemy.text(f"""DELETE FROM {models.LicenseServer.__tablename__};"""))
+        await sess.execute(sqlalchemy.text(f"""DELETE FROM {models.Configuration.__tablename__};"""))
+        await sess.execute(sqlalchemy.text(f"""DELETE FROM {models.Cluster.__tablename__};"""))
+        await sess.commit()
```
Comment (dusktreader) [inline backend/tests/conftest.py]: This is going to make your tests very slow. The way to handle this is to make the session during unit tests to flush their objects instead of commit. The database should be rolled back after each test function finishes.

**Review by dusktreader — CHANGES_REQUESTED** [review]

File: backend/Dockerfile (hunk 1 above)
Comment (julianaklulo) [reply backend/Dockerfile]: I created a separate Dockerfile for the CI and reverted this change.

File: backend/alembic/env.py (hunk 2 above)
Comment (julianaklulo) [reply backend/alembic/env.py]: The [SOS API](https://github.com/omnivector-solutions/sos/blob/c169acae6b62a037e31d4ee9103ea682b561b7e6/sos-api/alembic/env.py#L10) uses it, @matheushent said it should prevent issues when running the migrations inside the container.
I tested running it without this line and it worked, so I removed it.

File: backend/alembic/versions/0003--implement_new_data_struture_schema.py (hunk 3 above)
Comment (julianaklulo) [reply backend/alembic/versions/0003--implement_new_data_struture_schema.py]: I removed it.

File: backend/alembic/versions/0003--implement_new_data_struture_schema.py (hunk 4 above)
Comment (julianaklulo) [reply backend/alembic/versions/0003--implement_new_data_struture_schema.py]: I removed the downgrade commands and the old migrations, leaving only the new migration.

File: backend/docker-compose.yaml (hunk 5 above)
Comment (julianaklulo) [reply backend/docker-compose.yaml]: This also came from the [SOS API](https://github.com/omnivector-solutions/sos/blob/main/sos-api/docker-compose.yaml#L84), I updated it to use poetry, since these dependencies are already in the pyproject.toml.

File: backend/lm_backend/api/cruds/booking.py (hunk 7 above)
Comment (julianaklulo) [reply backend/lm_backend/api/cruds/booking.py]: I removed it.

File: backend/lm_backend/api/cruds/booking.py (hunk 9 above)
Comment (julianaklulo) [reply backend/lm_backend/api/cruds/booking.py]: I updated the docstring to include:

> To determine if a booking can be made, we check if the amount of licenses already booked, the licenses
>  in use, the licenses reserved and the amount requested, are smaller or equal the license total.

Does it represent well the inequality?

File: backend/lm_backend/api/cruds/generic.py (hunk 10 above)
Comment (julianaklulo) [reply backend/lm_backend/api/cruds/generic.py]: You're right, it's not needed so I removed it.

File: backend/lm_backend/api/cruds/generic.py (hunk 11 above)
Comment (julianaklulo) [reply backend/lm_backend/api/cruds/generic.py]: My bad, I fixed it.

File: backend/lm_backend/api/routes/bookings.py (hunk 15 above)
Comment (julianaklulo) [reply backend/lm_backend/api/routes/bookings.py:29]: We get request logging with FastAPI, not sure if we need more logging than what's already provided.

File: backend/lm_backend/api/routes/products.py (hunk 20 above)
Comment (julianaklulo) [reply backend/lm_backend/api/routes/products.py]: That's a great idea, I updated the GenericCRUD to return the model name for each message.
Also updated the delete method to return the confirmation message.

File: backend/lm_backend/api/schemas.py (hunk 22 above)
Comment (julianaklulo) [reply backend/lm_backend/api/schemas.py]: Done!

File: backend/tests/api/routes/conftest.py (hunk 26 above)
Comment (julianaklulo) [reply backend/tests/api/routes/conftest.py]: Leftovers 🤦‍♀️

File: backend/tests/api/routes/conftest.py (hunk 27 above)
Comment (julianaklulo) [reply backend/tests/api/routes/conftest.py]: Leftovers 🤦‍♀️

File: backend/lm_backend/api/models.py (hunk 13 above)
Comment (julianaklulo) [reply backend/lm_backend/api/models.py]: Good catch, there were some more missing. I fixed them all.

File: backend/lm_backend/api/routes/clusters.py (hunk 16 above)
Comment (julianaklulo) [reply backend/lm_backend/api/routes/clusters.py:104]: Fixed.

File: backend/lm_backend/api/routes/jobs.py (hunk 17 above)
Comment (julianaklulo) [reply backend/lm_backend/api/routes/jobs.py:82]: Fixed.

File: backend/lm_backend/api/routes/jobs.py (hunk 19 above)
Comment (julianaklulo) [reply backend/lm_backend/api/routes/jobs.py:108]: Fixed.

File: backend/tests/api/routes/test_clusters.py (hunk 28 above)
Comment (julianaklulo) [reply backend/tests/api/routes/test_clusters.py]: The name of the tests are self explanatory, I don't think we need doctrings for them. I removed these.

File: backend/lm_backend/api/routes/jobs.py (hunk 18 above)
Comment (julianaklulo) [reply backend/lm_backend/api/routes/jobs.py:94]: We also need the exception in case the job is not in the list. Raising in the end makes it cover both cases (empty list and job not in the list).

File: backend/lm_backend/api/schemas.py (hunk 23 above)
Comment (julianaklulo) [reply backend/lm_backend/api/schemas.py]: Totally, I changed it to use PositiveInt.

File: backend/lm_backend/database.py (hunk 25 above)
Comment (julianaklulo) [reply backend/lm_backend/database.py:19]: I was not sure how to get the dialect once I removed the databases from this file. Do we really need it? From my tests it rendered the queries correctly.

File: backend/tests/conftest.py (hunk 30 above)
Comment (julianaklulo) [reply backend/tests/conftest.py:52]: The test_app is used to instantiate the backend_client, which is then used by the tests.
We need the backend_client to inject the auth token, so invoking the app in each test wouldn't be feasible.

File: backend/lm_backend/api/models.py (hunk 12 above)
Comment (julianaklulo) [reply backend/lm_backend/api/models.py]: I broke this file in modules inside a `models/ `folder.

File: backend/lm_backend/api/schemas.py (hunk 21 above)
Comment (julianaklulo) [reply backend/lm_backend/api/schemas.py]: I broke it down into separate files.

File: backend/lm_backend/database.py (hunk 24 above)
Comment (julianaklulo) [reply backend/lm_backend/database.py:2]: We can do this change later, as it envolves changing the name in many places.

File: backend/tests/api/routes/test_bookings.py (hunk 29 above)
Comment (julianaklulo) [reply backend/tests/api/routes/test_bookings.py:176]: I added test cases for:
```
get by id, update, delete with bad parameter
update with bad data
```

File: backend/tests/conftest.py (hunk 31 above)
Comment (dusktreader) [reply backend/tests/conftest.py]: We may revisit this in the future, but resolving for now.

**Review by dusktreader — APPROVED** [review]
Fantastic work! Way to stick with it through the long haul.
